### PR TITLE
Massively reduced the number of code host API requests Sourcegraph performs

### PR DESCRIPTION
**Pull request by slimsag**
# TL;DR

Prior to this change, for every search result you viewed we would perform 4 code host API requests and this led to us running out of our rate limit rapidly. It causes:

- [Search results to take a very long time to load when using Bitbucket Server.](https://github.com/sourcegraph/sourcegraph/issues/2618)
- Links to the GitHub repository, etc. to not appear on repo pages.
- And likely many, many, more issues.

After this change, we massively reduce the number of code host API requests we perform because it turned out in 99% of cases (not just when searching) we were doing them for literally no reason due to unfiled technical debt.

**IMPORTANT:** This change **does NOT** change any user-facing behavior. It only prevents us from doing needless work. The behavior observed _in all cases_ before and after this change is _identical_.

# Detailed explanation

Repository revision resolutions (`ResolveRev` and `GetCommit`) occur very often in our codebase. For example, every search result you view makes two of these calls.

Prior to this change, each revision resolution would consume two code host API requests (which are heavily rate limited) due to the fact that they must perform a repo-updater `/repo-lookup` request in order to determine what the remote URL for the repository is when updating the gitserver repo mirror.

We already knew this was costly, and in fact that is why there is a `CachedGitRepo` variant of `GitRepo` in this same file. Basically, if you don't need the remote URL we can spare ourselves tons of work and, most importantly, code host API reservations.

But revision resolution wasn't able to use `CachedGitRepo` because we operated under the assumption that when resolving a revision you _do_ want the repository to update. For example, when resolving `master` or `mybranch` you want the repository updated to reflect the latest state of those branches.

Times changed, repo-updater changed how we did things, and we started operating with more explicit requests to update repositories via `repoupdater.DefaultClient.EnqueueRepoUpdate`.

At some point, gitserver became not responsible for updating repositories _unless_ the rev in question was not known to it (i.e. `mybranch` may be stale if already known to gitserver until repo-updater updates it, but gitserver will explicitly update the repository if `mybranch` is unknown to gitserver). You can see this logic clearly in gitserver's [`ensureRevision` function](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@e1f88f6dfc5cbdc6cac0754cf4eeff1492609253/-/blob/cmd/gitserver/server/server.go#L1275-1276).

What this means is that despite the fact that we always use a `GitRepo` for `ResolveRev` and `GetCommit` requests, in 99% of cases gitserver already has the rev and will not update it, _but we'll still ask repo-updater to do all of that work and make those code-host requests just cuz'._

After this change, the user observed behavior remains the exact same: gitserver doesn't update revs it is already aware of (but note our UI does ask it to in specific locations via `EnqueueRepoUpdate`). We just don't ask repo-updater to do all that work for nothing anymore.

Helps #2618

### CODE REVIEW

**Comment by slimsag** [comment]
(I won't merge this before doing more substantial testing to ensure this does fix #2618 , updating the changelog, etc. but I wanted to get this out tonight for my euro counterparts to take a look :) PTAL if you have time @keegancsmith @tsenart )

**Comment by slimsag** [comment]
@keegancsmith Actually, I think you misunderstood the code here: the implementation for CachedGitRepo shows it returns a repository without a remote URL in the event the repository is not a github.com/gitlab.com one. i.e., `CachedGitRepo` never performs a `RepoLookup` request, which is why my change here worked for all repositories.

However, after you saying this and pointing out the `remoteURLFunc` it became clear to me that implementing it that way (while it is much more code) is nicer because it is more explicit and clear behavior rather than somewhat implicit behavior. PTAL at my new attempt here :)

File: pkg/vcs/git/commits.go [hunk 1]
```diff
@@ -105,37 +113,86 @@ func isInvalidRevisionRangeError(output, obj string) bool {
 // commitLog returns a list of commits.
 //
 // The caller is responsible for doing checkSpecArgSafety on opt.Head and opt.Base.
-func commitLog(ctx context.Context, repo gitserver.Repo, opt CommitsOptions) ([]*Commit, error) {
+func commitLog(ctx context.Context, repo gitserver.Repo, remoteURLFunc func() (string, error), opt CommitsOptions) ([]*Commit, error) {
+	// TODO: use it
```
Comment (keegancsmith) [inline pkg/vcs/git/commits.go]: todo? looks like it is used.

File: pkg/vcs/git/commits.go [hunk 2]
```diff
@@ -43,14 +44,16 @@ type CommitsOptions struct {
 	After  string // include only commits after this date
 
 	Path string // only commits modifying the given path are selected (optional)
+
+	NoEnsureRevision bool // do not try to fetch from remote if revision doesn't exist locally
```
Comment (keegancsmith) [inline pkg/vcs/git/commits.go]: you could instead pass the remote url func into the opts.

File: pkg/vcs/git/commits.go [hunk 3]
```diff
@@ -105,37 +113,86 @@ func isInvalidRevisionRangeError(output, obj string) bool {
...
+
 	cmd := gitserver.DefaultClient.Command("git", args...)
 	cmd.Repo = repo
 	data, stderr, err := cmd.DividedOutput(ctx)
-	if err != nil {
+	if err == nil {
+		return parse(data, stderr)
+	}
+
+	tryAgain := func(err error) bool {
```
Comment (keegancsmith) [inline pkg/vcs/git/commits.go]: there is no point trying again if remoteURLFunc is nil

**Comment by keegancsmith** [comment]
So it is a bit hacky to copy paste this, but TBH its worth it for now. This all needs to be cleaned up anyways.

If you wanna make it a bit less hacky, you could have this sort of logic live in `git.Command`. IE you do something like `WithURLRemoteFunc(remoteURLFunc, NoEnsureRevision)` then git.Command can do the wrapping since it understands the not found error in its layer.

**Comment by beyang** [comment]
Quick high-level questions here:
* Will this change the end-user behavior of "if I hard refresh a repo page, it will enqueue an update"?
* I think ideally we'd want to have one `GitRepo` function that returns the git repo, by default cached but the cache is bypassable via an option param. Are we getting closer to that here?

**Comment by slimsag** [comment]
> Will this change the end-user behavior of "if I hard refresh a repo page, it will enqueue an update"?

No, this PR does not change any user-facing behavior.

> I think ideally we'd want to have one GitRepo function that returns the git repo, by default cached but the cache is bypassable via an option param. Are we getting closer to that here?

I agree, but not in the scope of this PR, I am not refactoring / working on that code here. The correct way to do this would be to remove all necessity of choosing between cached/uncached I think, and having our git code defer to repo-updater when needed, but that is a larger change.

**Comment by slimsag** [comment]
> If you wanna make it a bit less hacky, you could have this sort of logic live in git.Command. IE you do something like WithURLRemoteFunc(remoteURLFunc, NoEnsureRevision) then git.Command can do the wrapping since it understands the not found error in its layer.

I gave this a solid shot just now, but unfortunately it is much more difficult than anticipated.

Because every pkg/vcs/git.* command is essentially responsible for parsing revision-not-found errors and creating `&RevisionNotFoundError` types, it is not readily possible for us to implement this logic inside of e.g. `gitserver.DefaultClient.Command` (there is no `git.Command` as you referenced, I assumed that was typo-ish).

What I can do is create a higher level abstraction around `gitserver.DefaultClient.Command` which does this, so at least the logic is shared. I'm doing that now.

**Comment by keegancsmith** [comment]
BTW feel free to ship an earlier version of this PR to unblock the affected customer(s). The cleaner solution can always be a follow-up PR. Either way ping me on Slack for a faster turn around on review time when it is ready.

**Comment by slimsag** [comment]
@keegancsmith PTAL

My earlier versions of this PR as it turns out were suffering from a bug in a specific edge case, so I will instead merge this version now (to unblock affected customers). The implementation is roughly the same as the one you approved earlier, with the only exception being the implementation is now shared + a buggy edge case is now fixed. The code is pretty well tested, so I am not worried. I'll follow up on any comments / suggestions you have of course.
